src/bundles/csg/functions.ts

```diff
@@ -0,0 +1,823 @@
+}
+
+//TODO
+/**
+ * Center the provided shape with the middle base of the shape at (0, 0, 0).
+ *
+ * @param {Shape} shape - The shape to be centered
+ * @returns {Shape} The shape that is centered
+ */
+function shapeSetOrigin(shape: Shape) {
+  let newSolid: Solid = align({ modes: ['min', 'min', 'min'] }, shape.solid);
+  return new Shape(newSolid);
+}
+
+/**
+ * Checks if the specified argument is a Shape.
+ *
+ * @param {unknown} argument - The value to check.
+ * @returns {boolean} Whether the argument is a Shape.
+ */
+export function is_shape(argument: unknown): boolean {
+  return argument instanceof Shape;
+}
+
+/**
+ * Creates a clone of the specified Shape.
+ *
+ * @param {Shape} shape - The Shape to be cloned.
+ * @returns {Shape} The cloned Shape.
+ */
+export function clone(shape: Shape): Shape {
+  return shape.clone();
+}
+
+/**
+ * Stores a clone of the specified Shape for later rendering.
+ *
+ * @param {Shape} shape - The Shape to be stored.
+ */
+export function store(shape: Shape): void {
+  Core.getRenderGroupManager().storeShape(shape.clone());
+}
+
+/**
+ * Colours a clone of the specified Shape using the specified hex colour code,
+ * then stores it for later rendering. You may use one of the default colour
+ * variables provided by the module, or you may specify your own custom colour
+ * code.
+ *
+ * Colour codes must be of the form "#XXXXXX" or "XXXXXX", where each X
+ * represents a non-case sensitive hexadecimal number. Invalid colour codes
+ * default to black.
+ *
+ * @param {Shape} shape - The Shape to be coloured and stored.
+ * @param {string} hex - The colour code to use.
+ */
+export function store_as_color(shape: Shape, hex: string): void {
+  let color: Color = hexToColor(hex);
+  let coloredSolid: Solid = colorize(color, shape.solid);
+  Core.getRenderGroupManager().storeShape(new Shape(coloredSolid));
+}
+
+/**
+ * Colours a clone of the specified Shape using the specified RGB values, then
+ * stores it for later rendering.
+ *
+ * RGB values are clamped between 0 and 1.
+ *
+ * @param {Shape} shape - The Shape to be coloured and stored.
+ * @param {number} redComponent - The colour's red component.
+ * @param {number} greenComponent - The colour's green component.
+ * @param {number} blueComponent - The colour's blue component.
+ */
+export function store_as_rgb(
+  shape: Shape,
+  redComponent: number,
+  greenComponent: number,
+  blueComponent: number
+): void {
+  redComponent = clamp(redComponent, 0, 1);
+  greenComponent = clamp(greenComponent, 0, 1);
+  blueComponent = clamp(blueComponent, 0, 1);
+
+  let coloredSolid: Solid = colorize(
+    [redComponent, greenComponent, blueComponent],
+    shape.solid
+  );
+  Core.getRenderGroupManager().storeShape(new Shape(coloredSolid));
+}
+
+/**
+ * Renders using any Shapes stored thus far, along with a grid and axis. The
+ * Shapes will then not be included in any subsequent renders.
+ */
+export function render_grid_axis(): RenderGroup {
```

Comment:
It is challenging to have modules operate with native Source lists, for something like checking/specifying indexes of the shapes of a render group to actually render. As CSG is planned for use early into CS1101S, around the time of rune, I do not think arrays would be an option either. Such a suite of functions would increase complexity. It may be simpler just having users store the specific shapes they want before each render call (each render call clears all stored shapes for the next call, if any).

```js
store(my_cube);
store(my_sphere);
render();

store(my_cube);
store(my_prism);
render();
```

Or, if this is referring to something like a `is_stored(my_cube)`, that would probably require using a new property within each shape to remember stored status, as the object that gets stored is not that which gets passed in (especially when the shape is to be coloured before storage). All the passed shapes would need to be tracked separately, with any call to `render...`resetting their stored status. I could see potential uses for such a function as an enhancement, though it could also be left to the user to implement their own tracked storage abstractions, should they wish to avoid storing duplicate shapes in some kind of iteration logic

Reply:
I was thinking more along the lines of a list displayed in the tab? I understand that creating a string representation of the shapes might be difficult, but it would be helpful with debugging if the user could see what shapes are being displayed

Reply:
Could possibly experiment with something like [this](https://stackoverflow.com/a/42791996/11536796) wrt the naming, though I feel that this would be a more secondary feature